File: 907-KokoEatingBananas/907-KokoEatingBananas.py
# Last updated: 11/2/2025, 8:22:19 PM
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        
        left = 1  
        right = max(piles)
        res = right

        while left <= right:
            mid = left + (right-left)//2

            #see how many hours it takes to finish all the bananas at the current eating speed 
            totalTime = 0 
            for p in piles:
                totalTime += math.ceil(p/mid)
            
            #valid eating speed, see if there is a faster possible eating speed 
            if totalTime <= h:
                res = mid
                right = mid - 1  
            #not valid eating speed,eating too slow 
            else: #totalTime > h
                left = mid + 1

        return res 
            

File: 907-KokoEatingBananas/907-KokoEatingBananas.java
// Last updated: 11/2/2025, 8:22:20 PM
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        // Initalize the left and right boundaries 
        int left = 1;
        int right = 1;
        for (int pile : piles) {
            right = Math.max(right, pile);
        }

        while (left < right) {
            // Get the mid index between left and right boundary indexes.
            // hourSpent stands for the total hour Koko spends.
            int mid = (left + right) / 2;
            int hourSpent = 0;

            // Iterate over the piles and calculate hourSpent.
            // We increase the hourSpent by ceil(pile / mid)
            for (int pile : piles) {
                hourSpent += Math.ceil((double) pile / mid);
            }

            // Check if mid is a workable speed, and cut the search space by half.
            if (hourSpent <= h) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        // Once the left and right boundaries coincide, we find the target value,
        // that is, the minimum workable eating speed.
        return right;
    }
}